README.md
# timestring

[![Version](https://img.shields.io/npm/v/timestring.svg?style=flat-square)](https://www.npmjs.com/package/timestring)
[![Build Status](https://img.shields.io/github/workflow/status/mike182uk/timestring/CI/master?style=flat-square)](https://github.com/mike182uk/timestring/actions?query=workflow%3ACI)
[![Coveralls](https://img.shields.io/coveralls/mike182uk/timestring/master.svg?style=flat-square)](https://coveralls.io/r/mike182uk/timestring)
[![npm](https://img.shields.io/npm/dm/timestring.svg?style=flat-square)](https://www.npmjs.com/package/timestring)
[![License](https://img.shields.io/github/license/mike182uk/timestring.svg?style=flat-square)](https://www.npmjs.com/package/timestring)

Parse a human readable time string into a time based value.

## Installation

```bash
npm install --save timestring
```

## Usage

### Overview

```js
const timestring = require('@navy.gif/timestring')

let str = '1h 15m'
let time = timestring(str)

console.log(time) // will log 4500
```

**By default the returned time value from `timestring` will be in seconds.**

The time string can contain as many time groups as needed:

```js
const timestring = require('@navy.gif/timestring')

let str = '1d 3h 25m 18s'
let time = timestring(str)

console.log(time) // will log 98718
```

and can be as messy as you like:

```js
const timestring = require('@navy.gif/timestring')

let str = '1 d    3HOurS 25              min         1   8s'
let time = timestring(str)

console.log(time) // will log 98718
```

### Keywords

`timestring` will parse the following keywords into time values:

1. `ms, milli, millisecond, milliseconds` - will parse to milliseconds
2. `s, sec, secs, second, seconds` - will parse to seconds
3. `m, min, mins, minute, minutes` - will parse to minutes
4. `h, hr, hrs, hour, hours` - will parse to hours
5. `d, day, days` - will parse to days
6. `w, week, weeks` - will parse to weeks
7. `mo, mon, mth, mths, month, months` - will parse to months
8. `y, yr, yrs, year, years` - will parse to years

Keywords can be used interchangeably:

```js
const timestring = require('@navy.gif/timestring')

let str = '1day 15h 20minutes 15s'
let time = timestring(str)

console.log(time) // will log 141615
```

### Return Time Value

By default the return time value will be in seconds. This can be changed by passing one of the following strings as an argument to `timestring`:

1. `ms` - Milliseconds
2. `s` - Seconds
3. `m` - Minutes
4. `h` - Hours
5. `d` - Days
6. `w` - Weeks
7. `mth` - Months
8. `y` - Years

```js
const timestring = require('@navy.gif/timestring')

let str = '22h 16m'

let hours = timestring(str, 'h')
let days = timestring(str, 'd')
let weeks = timestring(str, 'w')

console.log(hours) // will log 22.266666666666666
console.log(days) // will log 0.9277777777777778
console.log(weeks) // will log 0.13253968253968254
```

### Optional Configuration

A few assumptions are made by default:

1. There are 24 hours per day
2. There are 7 days per week
3. There are 4 weeks per month
4. There are 12 months per year
5. There are 365.25 days per year

These options can be changed by passing an options object as an argument to `timestring`.

The following options are configurable:

1. `hoursPerDay`
2. `daysPerWeek`
3. `weeksPerMonth`
4. `monthsPerYear`
5. `daysPerYear`

```js
const timestring = require('@navy.gif/timestring')

let str = '1d'
let opts = {
  hoursPerDay: 1
}

let time = timestring(str, 'h', opts)

console.log(time) // will log 1
```

In the example above `hoursPerDay` is being set to `1`. When the time string is being parsed, the return value is being specified as hours. Normally `1d` would parse to `24` hours (as by default there are 24 hours in a day) but because `hoursPerDay` has been set to `1`, `1d` will now only parse to `1` hour.

This would be useful for specific application needs.

*Example - Employees of my company work 7.5 hours a day, and only work 5 days a week. In my time tracking app, when they type `1d` i want 7.5 hours to be tracked. When they type `1w` i want 5 days to be tracked etc.*

```js
const timestring = require('@navy.gif/timestring')

let opts = {
  hoursPerDay: 7.5,
  daysPerWeek: 5
}

let hoursToday = timestring('1d', 'h', opts)
let daysThisWeek = timestring('1w', 'd', opts)

console.log(hoursToday) // will log 7.5
console.log(daysThisWeek) // will log 5
```

It is important to note that the `daysPerYear` configuration option will be used to convert a month or year to seconds, so if you are using custom configuration options make sure that you adjust this value to suit if you expect to be parsing timestrings containing months or years.

## Notes

If the string that is passed into `timestring` can not be parsed then an error will be thrown:

```js
const timestring = require('@navy.gif/timestring')

let str = 'aaabbbccc'
let time = timestring(str) // will throw an error
```


CHANGELOG.md
# Changelog

## 6.0.2 - 6.0.6
Multiple version bumps because brain no worky

- TS typings

## 6.0.1

- Added `mo` as a shorthand for month

## 6.0.0

- Drop support for Node.js `< 8.0.0`
- Throw an error when unable to parse a timestring ([prolink007](https://github.com/prolink007))

## 5.0.1

- Fix issue where decimal values were not being parsed correctly ([vegeta897](https://github.com/vegeta897))

## 5.0.0

- Add `daysPerYear` configuration option
- Use `daysPerYear` configuration option to convert months or years to seconds

## 4.0.0

- Drop support for Node.js `< 4.0.0`
- Remove dependency on `lodash`

## 3.2.0

- Add `mon` unit ([@IanMitchell](https://github.com/IanMitchell))

## 3.1.0

- Add `ms` unit ([@mrbar42](https://github.com/mrbar42))

## 3.0.0

- Remove `String.parseTime`
- Export function instead of object

## 2.0.0

- Keep it simple: revert back to ES5, remove gulp, browserify and build steps
- Remove bower / browser specific integrations - This module can still be used client side using modern development tools like webpack, browserify etc.
- use `node-style-guide` coding style

## 1.1.1

- src now uses ES6
- use gulp instead of grunt
- setup gulp build pipeline
- use analysis tools: JSCS, jshint
- use consistent coding style (jQuery coding style)
- send code coverage to coveralls.io
- use browserify to generate distributable scripts
- add contributing doc
- misc meta data updates
- add sourcemaps

## 1.1.0

- add MIT license
- add Changelog
- add Grunt config
- add editorconfig
- add Tests
- fix JSHint errors
- format code as per editorconfig
- fix bug with conversion from year(s) to another unit
- make sure `Error` (type) is thrown when invalid unit is encountered
- add minified distributable version
- update package.json
- add bower config
- add travis config


src/index.js
/**
 * Exports
 */

module.exports = parseTimestring

/**
 * Default options to use when parsing a timestring
 *
 * @type {Object}
 */

const DEFAULT_OPTS = {
  hoursPerDay: 24,
  daysPerWeek: 7,
  weeksPerMonth: 4,
  monthsPerYear: 12,
  daysPerYear: 365.25
}

/**
 * Map of accepted strings to unit
 *
 * @type {Object}
 */

const UNIT_MAP = {
  ms: ['ms', 'milli', 'millisecond', 'milliseconds'],
  s: ['s', 'sec', 'secs', 'second', 'seconds'],
  m: ['m', 'min', 'mins', 'minute', 'minutes'],
  h: ['h', 'hr', 'hrs', 'hour', 'hours'],
  d: ['d', 'day', 'days'],
  w: ['w', 'week', 'weeks'],
  mth: ['mo', 'mon', 'mth', 'mths', 'month', 'months'],
  y: ['y', 'yr', 'yrs', 'year', 'years']
}

/**
 * Parse a timestring
 *
 * @param   {string} string
 * @param   {string} returnUnit
 * @param   {Object} opts
 * @returns {number}
 */

function parseTimestring (string, returnUnit, opts) {
  opts = Object.assign({}, DEFAULT_OPTS, opts || {})

  let totalSeconds = 0
  const unitValues = getUnitValues(opts)
  const groups = string
    .toLowerCase()
    .replace(/[^.\w+-]+/g, '')
    .match(/[-+]?[0-9.]+[a-z]+/g)

  if (groups === null) {
    throw new Error(`The string [${string}] could not be parsed by timestring`)
  }

  groups.forEach(group => {
    const value = group.match(/[0-9.]+/g)[0]
    const unit = group.match(/[a-z]+/g)[0]

    totalSeconds += getSeconds(value, unit, unitValues)
  })

  if (returnUnit) {
    return convert(totalSeconds, returnUnit, unitValues)
  }

  return totalSeconds
}

/**
 * Get unit values based on the passed options
 *
 * @param   {Object} opts
 * @returns {Object}
 */

function getUnitValues (opts) {
  const unitValues = {
    ms: 0.001,
    s: 1,
    m: 60,
    h: 3600
  }

  unitValues.d = opts.hoursPerDay * unitValues.h
  unitValues.w = opts.daysPerWeek * unitValues.d
  unitValues.mth = (opts.daysPerYear / opts.monthsPerYear) * unitValues.d
  unitValues.y = opts.daysPerYear * unitValues.d

  return unitValues
}

/**
 * Get the key for a unit
 *
 * @param   {string} unit
 * @returns {string}
 */

function getUnitKey (unit) {
  for (const key of Object.keys(UNIT_MAP)) {
    if (UNIT_MAP[key].indexOf(unit) > -1) {
      return key
    }
  }

  throw new Error(`The unit [${unit}] is not supported by timestring`)
}

/**
 *  Get the number of seconds for a value, based on the unit
 *
 * @param   {number} value
 * @param   {string} unit
 * @param   {Object} unitValues
 * @returns {number}
 */

function getSeconds (value, unit, unitValues) {
  return value * unitValues[getUnitKey(unit)]
}

/**
 * Convert a value from its existing unit to a new unit
 *
 * @param   {number} value
 * @param   {string} unit
 * @param   {Object} unitValues
 * @returns {number}
 */

function convert (value, unit, unitValues) {
  return value / unitValues[getUnitKey(unit)]
}
